volumes.cpp

```diff
@@ -163,6 +163,258 @@ class PlaneSurface : public SurfaceVolume {
 
 typedef  enum { X=0, Y=1, Z=2 } axis_t;
 
+class GeneralQuadraticSurface : public SurfaceVolume {
+
+protected:
+  double A_,B_,C_,D_,E_,F_,G_,H_,J_,K_;
+  int type;
+  Vector3d translation;
+  double rotation_mat[9];
+  double extents[3];
+  double gq_tol = 1e-6;
+
+  enum GQ_TYPE {UNKNOWN = 0,
+		ELLIPSOID,
+		ONE_SHEET_HYPERBOLOID,
+		TWO_SHEET_HYPERBOLOID,
+		ELLIPTIC_CONE,
+		ELLIPTIC_PARABOLOID,
+		HYPERBOLIC_PARABOLOID,
+		ELLIPTIC_CYL,
+		HYPERBOLIC_CYL,
+		PARABOLIC_CYL};
+public:
+  GeneralQuadraticSurface(double A, double B, double C, double D, double E, double F, double G, double H, double J, double K):
+    SurfaceVolume(),A_(A),B_(B),C_(C),D_(D),E_(E),F_(F),G_(G),H_(H),J_(J),K_(K)
+  { cannonize(); }
+
+  virtual double getFarthestExtentFromOrigin() const{ return 0; }
+protected:
+  void cannonize()
```

Comment:
Language nerdery: "Cannonize" isn't a word, but if it was, it would mean something like "to put through a cannon."  "Canonize" means to declare sainthood, or to glorify.  If what you wanted was a word meaning "to put into a canonical form," then the expected name would be "canonicalize", ugly as that is.